snuba/subscriptions/entity_subscription.py

```diff
@@ -0,0 +1,122 @@
+from abc import ABC, abstractmethod
+from enum import Enum
+from typing import Any, List, Mapping, Optional, Sequence, Type
+
+from snuba.datasets.entities import EntityKey
+from snuba.query.conditions import ConditionFunctions, binary_condition
+from snuba.query.exceptions import InvalidQueryException
+from snuba.query.expressions import Column, Expression, FunctionCall, Literal
+from snuba.query.types import Condition
+
+
+class SubscriptionType(Enum):
+    LEGACY = "legacy"
+    SNQL = "snql"
+    DELEGATE = "delegate"
+
+
+class EntitySubscription(ABC):
+    def __init__(self, data_dict: Mapping[str, Any]) -> None:
+        ...
+
+    @abstractmethod
+    def get_entity_subscription_conditions_for_snql(
+        self, offset: Optional[int] = None
+    ) -> List[Expression]:
+        """
+        Returns a list of extra conditions that are entity specific and required for the
+        snql subscriptions
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def get_entity_subscription_conditions_for_legacy(
+        self, offset: Optional[int] = None
+    ) -> Sequence[Condition]:
+        """
+        Returns a list of extra conditions that are entity specific and required for the
+        legacy subscriptions
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def to_dict(self) -> Mapping[str, Any]:
+        raise NotImplementedError
+
+
+class SessionsSubscription(EntitySubscription):
+    def __init__(self, data_dict: Mapping[str, Any]) -> None:
+        super().__init__(data_dict)
+        try:
+            self.organization: int = data_dict["organization"]
+        except KeyError:
+            raise InvalidQueryException(
+                "organization param is required for any query over sessions entity"
+            )
+
+    def get_entity_subscription_conditions_for_snql(
+        self, offset: Optional[int] = None
+    ) -> List[Expression]:
+        return [
+            binary_condition(
+                ConditionFunctions.EQ,
+                Column(None, None, "org_id"),
+                Literal(None, self.organization),
+            ),
+        ]
+
+    def get_entity_subscription_conditions_for_legacy(
+        self, offset: Optional[int] = None
+    ) -> Sequence[Condition]:
+        return []
+
+    def to_dict(self) -> Mapping[str, Any]:
+        return {"organization": self.organization}
```

Comment:
Where does this key `organization` end up getting used? I noticed we use the name `project_id` for the project ID. Should we follow a convention?

Reply:
It will fail because I think there is still an extension there that validates on `organization`

Reply:
What does this have to do with organization extensions though? Isn't it just the key used for encoding and decoding the subscription data in redis?

Reply:
Not really because there is extension validation that runs and validates against the `organization` param in Session entity right over here -> https://github.com/getsentry/snuba/blob/master/snuba/query/organization_extension.py#L8